src/dividat-driver/flex/main.go

```diff
@@ -0,0 +1,262 @@
+
+	for _, port := range ports {
+		// Terminate if we have been cancelled
+		if ctx.Err() != nil {
+			return
+		}
+
+		logger.WithField("name", port.Name).WithField("vendor", port.VID).Debug("Considering serial port.")
+
+		if isFlexLike(port) {
+			connectSerial(ctx, logger, port.Name, onReceive)
+		}
+	}
+}
+
+// Check whether a port looks like a potential Flex device.
+//
+// Vendor IDs:
+//   16C0 - Van Ooijen Technische Informatica (Teensy)
+func isFlexLike(port *enumerator.PortDetails) bool {
+	vendorId := strings.ToUpper(port.VID)
+
+	return vendorId == "16C0"
+}
+
+
+// Serial communication
+
+type ReaderState int
+
+const (
+	WAITING_FOR_HEADER ReaderState = iota
+	HEADER_START
+	HEADER_READ_LENGTH_MSB
+	WAITING_FOR_BODY
+	BODY_START
+	BODY_POINT
+	UNEXPECTED_BYTE
+)
+
+const (
+	HEADER_START_MARKER = 'N'
+	BODY_START_MARKER = 'P'
+)
+
+// Actually attempt to connect to an individual serial port and pipe its signal into the callback, summarizing
+// package units into a buffer.
+func connectSerial(ctx context.Context, logger *logrus.Entry, serialName string, onReceive func([]byte)) {
+	mode := &serial.Mode{
+		BaudRate: 115200,
+		Parity:   serial.NoParity,
+		DataBits: 8,
+		StopBits: serial.OneStopBit,
+	}
+
+	START_MEASUREMENT_CMD := []byte{'S', '\n'}
+
+	logger.WithField("name", serialName).Info("Attempting to connect with serial port.")
+	port, err := serial.Open(serialName, mode)
+	if err != nil {
+		logger.WithField("config", mode).WithField("error", err).Info("Failed to open connection to serial port.")
+		return
+	}
+	defer port.Close()
+
+	_, err = port.Write(START_MEASUREMENT_CMD)
+	if err != nil {
+		logger.WithField("error", err).Info("Failed to write start message to serial port.")
+		port.Close()
+		return
+	}
+
+	reader := bufio.NewReader(port)
+	state := WAITING_FOR_HEADER
+	pointsLeftInSet := 0
+	bytesLeftInPoint := 0
+
+	var buff []byte
+	for {
+		// Terminate if we were cancelled
+		if ctx.Err() != nil {
+			logger.WithField("name", serialName).Info("Disconnecting from serial port.")
+			return
+		}
+
+		input, err := reader.ReadByte()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			continue
+		}
+
+		// Finite State Machine for parsing byte stream
+		switch {
+		case state == WAITING_FOR_HEADER && input == HEADER_START_MARKER:
+			state = HEADER_START
+		case state == HEADER_START && input == '\n':
+			state = HEADER_READ_LENGTH_MSB
+                case state == HEADER_READ_LENGTH_MSB:
+			// The number of measurements in each set may vary and is
+			// given as two consecutive bytes (big-endian).
+			msb := input
+			lsb, err := reader.ReadByte()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				continue
+			}
+			pointsLeftInSet = int(binary.BigEndian.Uint16([]byte{msb,lsb}))
+			state = WAITING_FOR_BODY
+		case state == WAITING_FOR_BODY && input == BODY_START_MARKER:
+			state = BODY_START
+		case state == BODY_START && input == '\n':
+			state = BODY_POINT
+			bytesLeftInPoint = 4
```

Comment:
Use a constant for 4 ? It could also be used below, when starting the next point and reinitializing the variable.

Reply:
Also reset `buff` here ?

Reply:
Good idea, introduced a constant in c7902cc.

Not sure about resetting `buff`. Just for good measure? I think this shouldn't be needed unless there are mistaken writes to it elsewhere.